crates/xline-client/src/types/kv.rs

```diff
@@ -109,47 +111,36 @@ impl From<PutOptions> for xlineapi::PutRequest {
 }
 
 /// Request type for `Range`
-#[derive(Debug, PartialEq)]
-pub struct RangeRequest {
+#[derive(Debug, PartialEq, Default)]
+pub struct RangeOptions {
     /// Inner request
     inner: xlineapi::RangeRequest,
+    /// Range end options
+    range_end_options: RangeEndOptions,
 }
 
-impl RangeRequest {
-    /// Creates a new `RangeRequest`
-    ///
+impl RangeOptions {
     /// `key` is the first key for the range. If `range_end` is not given, the request only looks up key.
     #[inline]
-    pub fn new(key: impl Into<Vec<u8>>) -> Self {
-        Self {
-            inner: xlineapi::RangeRequest {
-                key: key.into(),
-                ..Default::default()
-            },
-        }
+    #[must_use]
+    pub fn with_key(mut self, key: impl Into<Vec<u8>>) -> Self {
+        self.inner.key = key.into();
+        self
     }
 
     /// If set, Xline will return all keys with the matching prefix
     #[inline]
     #[must_use]
     pub fn with_prefix(mut self) -> Self {
-        if self.inner.key.is_empty() {
-            self.inner.key = vec![0];
-            self.inner.range_end = vec![0];
-        } else {
-            self.inner.range_end = KeyRange::get_prefix(&self.inner.key);
-        }
+        self.range_end_options = RangeEndOptions::Prefix;
```

Comment:
theoretically, the three functions `with_prefix`, `with_from_key`, `with_range_end` should not be called more than once, because they are all about to set the `range_end`. The new implementation makes the last setting win, I think it's in expect.

Reply:
I mean calling this `range_end` method below, not `with_range_end`.
```rust
/// Get `range_end`
#[inline]
#[must_use]
pub fn range_end(&self) -> &[u8] {
    &self.inner.range_end
}
```

Reply:
ok, that's a problem. I think we can modify it to `range_end_option()` that returns a `RangeEndOption`, and remove the `key(&self)` and `range_end(&self)`, because these two functions `RangeRequest::range_end(&self)` and `RangeRequest::key(&self)` are actually not used in other place.

*revert these with... methods back to their original form* may cause inconsistency api between KvClient::put and other apis, and this also means we should revert `AuthRoleGrantPermissionRequest`, `Permission`, `DeleteRangeRequest`, `WatchRequest` because they has similar RangeEndOptions?

Reply:
`RangeRequest::key(&self)` is not affected by your change, so we don't need to remove it.

Since `RangeRequest::range_end(&self)` is a public API, I think we should not remove it. Can you solve the bug in a way that you prefer?

Reply:
> `RangeRequest::key(&self)` is not affected by your change, so we don't need to remove it.

after refactor, there'll be no `types::kv::RangeRequest`, that's the biggest breaking change, and we can put api compatibility aside. `RangeOptions` has `inner: xlineapi::RangeRequest`, there are `key` and `range_end` in it, but these value will keep default until KvClient calls `range(key, option)`, so I think calling a `key(&self)` or `range_end(&self)` on `RangeOptions` to get a default value is meaningless.

Reply:
I got it. Let's modify it as you suggested.